src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/KeyReadsByRefShardUnitTest.java

```diff
@@ -53,4 +55,56 @@ public void groupReadsForRefTest(List<GATKRead> reads, List<KV<ReferenceShard, I
 
         p.run();
     }
+
+    @DataProvider(name = "KeyReadsByRefShardWithCustomWindowFunctionTestData")
+    public Object[][] keyReadsByRefShardWithCustomWindowFunctionTestData() {
+        final List<GATKRead> samReads = ReadsPreprocessingPipelineTestData.makeReferenceShardBoundaryReads(2, 3, SAMRecord.class);
+        final List<GATKRead> googleReads = ReadsPreprocessingPipelineTestData.makeReferenceShardBoundaryReads(2, 3, Read.class);
+
+        return new Object[][] {
+                // Test case layout: reads, reference window function to apply, expected shard + reads pairs
+                { samReads, RefWindowFunctions.IDENTITY_FUNCTION, generateExpectedCustomWindowResult(samReads, 0, 0) },
+                { googleReads, RefWindowFunctions.IDENTITY_FUNCTION, generateExpectedCustomWindowResult(googleReads, 0, 0) },
+                { samReads, new RefWindowFunctions.FixedWindowFunction(1, 1), generateExpectedCustomWindowResult(samReads, 1, 1) },
+                { googleReads, new RefWindowFunctions.FixedWindowFunction(1, 1), generateExpectedCustomWindowResult(googleReads, 1, 1) },
+                { samReads, new RefWindowFunctions.FixedWindowFunction(3, 5), generateExpectedCustomWindowResult(samReads, 3, 5) },
+                { googleReads, new RefWindowFunctions.FixedWindowFunction(3, 5), generateExpectedCustomWindowResult(googleReads, 3, 5) },
+        };
+    }
+
+    private List<KV<ReferenceShard, Iterable<GATKRead>>> generateExpectedCustomWindowResult( final List<GATKRead> reads, final int windowLeadingBases, final int windowTrailingBases ) {
+        final Map<ReferenceShard, List<GATKRead>> shardMap = new HashMap<>();
+
+        for ( final GATKRead read : reads ) {
+            final SimpleInterval expandedReadInterval = new SimpleInterval(read.getContig(), Math.max(read.getStart() - windowLeadingBases, 1), read.getEnd() + windowTrailingBases);
+            final ReferenceShard readShard = ReferenceShard.getShardNumberFromInterval(expandedReadInterval);
+
+            if ( shardMap.containsKey(readShard) ) {
```

Comment:
Consider using Guava's Multimap. You should be able to avoid this if/else here.
https://code.google.com/p/guava-libraries/wiki/NewCollectionTypesExplained